VPLAY-10605 [VIPA] CC is not displayed in VOD assets with pre-roll ads

anjali-syna opened the pull request:
Reason for change: CC is not displayed in VOD assets as dynamic configuration of gstreamer subtitle pipeline is not possible. Adding work around in aamp to overcome this.
Changes:
* Adding code in aamp to create subtitle sink along with video and audio when the pipeline is created initially. Also fetch and inject the subtitle init fragments for successful creation of subtitle pipeline
Test Procedure: Refer JIRA ticket

anjali-syna committed:
VPLAY-10605 [VIPA] CC is not displayed in VOD assets with pre-roll ads

Reason for change: CC is not displayed in VOD assets as dynamic configuration
of gstreamer subtitle pipeline is not possible. Adding work around in aamp.

Changes:
* Adding code in aamp to create subtitle sink along with video and audio
  when the pipeline is created initially. Also fetch and inject the subtitle
  init fragments for successful creation of subtitle pipeline

Test Procedure: Refer JIRA ticket

Risks: medium

anjali-syna committed:
VPLAY-10605 [VIPA] CC is not displayed in VOD assets with pre-roll ads

Reason for change: CC is not displayed in VOD assets as dynamic configuration
of gstreamer subtitle pipeline is not possible. Adding work around in aamp.

Changes:
* Adding code in aamp to create subtitle sink along with video and audio
  when the pipeline is created initially. Also fetch and inject the subtitle
  init fragments for successful creation of subtitle pipeline

Test Procedure: Refer JIRA ticket

Risks: medium

rdkcmf-jenkins commented on fragmentcollector_mpd.cpp:
```diff
@@ -8956,6 +8961,104 @@
+										
+										ConstructFragmentURL(fragmentUrl,fragmentDescriptorCMCD , initialization);
+										AAMPLOG_MIL("ANJ: [SUBTITLE]: mimeType:%s, init url %s", subtitleMimeType.c_str(), fragmentUrl.c_str());
+										subtitleHeader->url = fragmentUrl;
```
<!-- Comment managed by coverity-report-output-v7 action, do not modify!
0139b294a05f11fa59e0a6b67642fd1c
NOT_PRESENT
-->

Coverity issue no longer present as of: undefined
<details>
<summary>Show issue</summary>


# Coverity Issue - Variable copied when it could be moved
"fragmentUrl" is copied in call to copy assignment for class "std::string", when it could be moved instead.

_Low Impact, CWE-none_
_COPY_INSTEAD_OF_MOVE_

## How to fix
 Use "std::move(""fragmentUrl"")" instead of "fragmentUrl".

</details>

rdkcmf-jenkins commented on fragmentcollector_mpd.cpp:
```diff
@@ -8956,6 +8961,104 @@
+										subtitleHeader->url = fragmentUrl;
+										subtitleHeader->mimeType = subtitleMimeType;
+										//AampStreamSinkManager::GetInstance().AddMediaHeader(aamp, eMEDIATYPE_SUBTITLE, subtitleHeader);
+										AampStreamSinkManager::GetInstance().AddMediaHeader(eMEDIATYPE_SUBTITLE, subtitleHeader);
```
<!-- Comment managed by coverity-report-output-v7 action, do not modify!
428cd2888bb3237e97243a24b140edcb
NOT_PRESENT
-->

Coverity issue no longer present as of: undefined
<details>
<summary>Show issue</summary>


# Coverity Issue - Variable copied when it could be moved
"subtitleHeader" is passed-by-value as parameter to "std::shared_ptr<AampStreamSinkManager::MediaHeader>::shared_ptr(std::shared_ptr<AampStreamSinkManager::MediaHeader> const &) /*explicit =default*/", when it could be moved instead.

_Low Impact, CWE-none_
_COPY_INSTEAD_OF_MOVE_

## How to fix
 Use "std::move(""subtitleHeader"")" instead of "subtitleHeader".

</details>

rdkcmf-jenkins commented on AampStreamSinkManager.cpp:
```diff
@@ -616,3 +616,57 @@ void AampStreamSinkManager::UpdateTuningPlayer(PrivateInstanceAAMP *aamp)
+	}
+	else
+	{
+		mMediaHeaders[track] = header;
```
<!-- Comment managed by coverity-report-output-v7 action, do not modify!
22d238af668211b03dde3ed67802de61
NOT_PRESENT
-->

Coverity issue no longer present as of: undefined
<details>
<summary>Show issue</summary>


# Coverity Issue - Variable copied when it could be moved
"header" is copied in call to copy assignment for class "std::shared_ptr<AampStreamSinkManager::MediaHeader>", when it could be moved instead.

_Low Impact, CWE-none_
_COPY_INSTEAD_OF_MOVE_

## How to fix
 Use "std::move(""header"")" instead of "header".

</details>

rdkcmf-jenkins commented on fragmentcollector_mpd.cpp:
```diff
@@ -8956,6 +8961,104 @@
+										ConstructFragmentURL(fragmentUrl,fragmentDescriptorCMCD , initialization);
+										AAMPLOG_MIL("ANJ: [SUBTITLE]: mimeType:%s, init url %s", subtitleMimeType.c_str(), fragmentUrl.c_str());
+										subtitleHeader->url = fragmentUrl;
+										subtitleHeader->mimeType = subtitleMimeType;
```
<!-- Comment managed by coverity-report-output-v7 action, do not modify!
5dfb2b091d9dc6d04423fc53feb7e300
NOT_PRESENT
-->

Coverity issue no longer present as of: undefined
<details>
<summary>Show issue</summary>


# Coverity Issue - Variable copied when it could be moved
"subtitleMimeType" is copied in call to copy assignment for class "std::string", when it could be moved instead.

_Low Impact, CWE-none_
_COPY_INSTEAD_OF_MOVE_

## How to fix
 Use "std::move(""subtitleMimeType"")" instead of "subtitleMimeType".

</details>

rdkcmf-jenkins commented on fragmentcollector_mpd.cpp:
```diff
@@ -8956,6 +8961,104 @@
+										fragmentDescriptor->RepresentationID.assign(representation->GetId());
+										FragmentDescriptor *fragmentDescriptorCMCD(fragmentDescriptor);
+										
+										ConstructFragmentURL(fragmentUrl,fragmentDescriptorCMCD , initialization);
```
<!-- Comment managed by coverity-report-output-v7 action, do not modify!
cbff14072be5a6b2a44e036698f04f48
NOT_PRESENT
-->

Coverity issue no longer present as of: undefined
<details>
<summary>Show issue</summary>


# Coverity Issue - Variable copied when it could be moved
"initialization" is passed-by-value as parameter to "std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::basic_string(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const &)", when it could be moved instead.

_Low Impact, CWE-none_
_COPY_INSTEAD_OF_MOVE_

## How to fix
 Use "std::move(""initialization"")" instead of "initialization".

</details>

rdkcmf-jenkins commented on fragmentcollector_mpd.cpp:
```diff
@@ -10834,6 +10939,23 @@
+	subtitleHeader = AampStreamSinkManager::GetInstance().GetMediaHeader(eMEDIATYPE_SUBTITLE);
+	if(subtitleHeader)
+	{
+		auto mimeType = subtitleHeader->mimeType;
```
<!-- Comment managed by coverity-report-output-v7 action, do not modify!
4ff94b445a8e809b6c318bbdddbf9eba
NOT_PRESENT
-->

Coverity issue no longer present as of: undefined
<details>
<summary>Show issue</summary>


# Coverity Issue - Use of auto that causes a copy
Using the "auto" keyword without an "&" causes the copy of an object of type "std::string".

_Low Impact, CWE-none_
_AUTO_CAUSES_COPY_

## How to fix
 Use "const auto &" instead of "auto".

</details>

anjali-syna committed:
Code cleaned up and comments added

Signed-off-by: anjali-syna <206662904+anjali-syna@users.noreply.github.com>

anjali-syna committed:
Code cleaned up and comments added

Signed-off-by: anjali-syna <206662904+anjali-syna@users.noreply.github.com>

rdkcmf-jenkins commented on fragmentcollector_mpd.cpp:
```diff
@@ -8956,6 +8962,91 @@
+
+								ConstructFragmentURL(fragmentUrl,fragmentDescriptorCMCD , initialization);
+								AAMPLOG_MIL("[SUBTITLE]: mimeType:%s, init url %s", subtitleMimeType.c_str(), fragmentUrl.c_str());
+								subtitleHeader->url = fragmentUrl;
```
<!-- Comment managed by coverity-report-output-v7 action, do not modify!
0139b294a05f11fa59e0a6b67642fd1c
NOT_PRESENT
-->

Coverity issue no longer present as of: undefined
<details>
<summary>Show issue</summary>


# Coverity Issue - Variable copied when it could be moved
"fragmentUrl" is copied in call to copy assignment for class "std::string", when it could be moved instead.

_Low Impact, CWE-none_
_COPY_INSTEAD_OF_MOVE_

## How to fix
 Use "std::move(""fragmentUrl"")" instead of "fragmentUrl".

</details>

rdkcmf-jenkins commented on fragmentcollector_mpd.cpp:
```diff
@@ -8956,6 +8962,91 @@
+								AAMPLOG_MIL("[SUBTITLE]: mimeType:%s, init url %s", subtitleMimeType.c_str(), fragmentUrl.c_str());
+								subtitleHeader->url = fragmentUrl;
+								subtitleHeader->mimeType = subtitleMimeType;
+								AampStreamSinkManager::GetInstance().AddMediaHeader(eMEDIATYPE_SUBTITLE, subtitleHeader);
```
<!-- Comment managed by coverity-report-output-v7 action, do not modify!
428cd2888bb3237e97243a24b140edcb
NOT_PRESENT
-->

Coverity issue no longer present as of: undefined
<details>
<summary>Show issue</summary>


# Coverity Issue - Variable copied when it could be moved
"subtitleHeader" is passed-by-value as parameter to "std::shared_ptr<AampStreamSinkManager::MediaHeader>::shared_ptr(std::shared_ptr<AampStreamSinkManager::MediaHeader> const &) /*explicit =default*/", when it could be moved instead.

_Low Impact, CWE-none_
_COPY_INSTEAD_OF_MOVE_

## How to fix
 Use "std::move(""subtitleHeader"")" instead of "subtitleHeader".

</details>

p-bond replied:
Ownership of subtitleHeader has been passed to AddMediaHeader, then it goes out of scope so this warning is correct.

anjali-syna replied:
done

rdkcmf-jenkins commented on fragmentcollector_mpd.cpp:
```diff
@@ -8956,6 +8962,91 @@
+								ConstructFragmentURL(fragmentUrl,fragmentDescriptorCMCD , initialization);
+								AAMPLOG_MIL("[SUBTITLE]: mimeType:%s, init url %s", subtitleMimeType.c_str(), fragmentUrl.c_str());
+								subtitleHeader->url = fragmentUrl;
+								subtitleHeader->mimeType = subtitleMimeType;
```
<!-- Comment managed by coverity-report-output-v7 action, do not modify!
5dfb2b091d9dc6d04423fc53feb7e300
NOT_PRESENT
-->

Coverity issue no longer present as of: undefined
<details>
<summary>Show issue</summary>


# Coverity Issue - Variable copied when it could be moved
"subtitleMimeType" is copied in call to copy assignment for class "std::string", when it could be moved instead.

_Low Impact, CWE-none_
_COPY_INSTEAD_OF_MOVE_

## How to fix
 Use "std::move(""subtitleMimeType"")" instead of "subtitleMimeType".

</details>

anjali-syna replied:
done

rdkcmf-jenkins commented on fragmentcollector_mpd.cpp:
```diff
@@ -8956,6 +8962,91 @@
+								fragmentDescriptor->RepresentationID.assign(representation->GetId());
+								FragmentDescriptor *fragmentDescriptorCMCD(fragmentDescriptor);
+
+								ConstructFragmentURL(fragmentUrl,fragmentDescriptorCMCD , initialization);
```
<!-- Comment managed by coverity-report-output-v7 action, do not modify!
cbff14072be5a6b2a44e036698f04f48
NOT_PRESENT
-->

Coverity issue no longer present as of: undefined
<details>
<summary>Show issue</summary>


# Coverity Issue - Variable copied when it could be moved
"initialization" is passed-by-value as parameter to "std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::basic_string(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const &)", when it could be moved instead.

_Low Impact, CWE-none_
_COPY_INSTEAD_OF_MOVE_

## How to fix
 Use "std::move(""initialization"")" instead of "initialization".

</details>

anjali-syna replied:
done

anjali-syna committed:
Added check for rialto and IsLiveStream

Signed-off-by: anjali-syna <206662904+anjali-syna@users.noreply.github.com>

anjali-syna committed:
Added check for rialto and IsLiveStream

Signed-off-by: anjali-syna <206662904+anjali-syna@users.noreply.github.com>

anjali-syna committed:
Updating mocks and Fakes

Signed-off-by: anjali-syna <206662904+anjali-syna@users.noreply.github.com>

anjali-syna committed:
Updating mocks and Fakes

Signed-off-by: anjali-syna <206662904+anjali-syna@users.noreply.github.com>

anjali-syna committed:
Merge branch 'dev_sprint_25_2' into feature/VPLAY-10605

anshephe committed:
Moved code to inject cached init segment to new function
Adjusted code obtaining the subtitleOutputFormat
Removed the unused method RemoveHeaders from StreamSinkManager
Added Rialto config to some tests setup

rdkcmf-jenkins commented on fragmentcollector_mpd.cpp:
```diff
@@ -11531,12 +11648,49 @@
+
+				// Update the header with injected set
+				header->injected = true;
+				AampStreamSinkManager::GetInstance().AddMediaHeader(iTrack, header);
```
<!-- Comment managed by coverity-report-output-v7 action, do not modify!
7880749200654e1e9c658767a3cce23b
NOT_PRESENT
-->

Coverity issue no longer present as of: undefined
<details>
<summary>Show issue</summary>


# Coverity Issue - Variable copied when it could be moved
"header" is passed-by-value as parameter to "std::shared_ptr<AampStreamSinkManager::MediaHeader>::shared_ptr(std::shared_ptr<AampStreamSinkManager::MediaHeader> const &) /*explicit =default*/", when it could be moved instead.

_Low Impact, CWE-none_
_COPY_INSTEAD_OF_MOVE_

## How to fix
 Use "std::move(""header"")" instead of "header".

</details>

anjali-syna replied:
done

anshephe committed:
Put temporary change to log functions back to original

anshephe committed:
Correction to change made to StreamAbstractionAAMP_MPD::GetStreamFormat

anjali-syna committed:
Merge branch 'dev_sprint_25_2' into feature/VPLAY-10605

anjali-syna committed:
Merge branch 'feature/VPLAY-10605' of github.com:rdkcentral/aamp into feature/VPLAY-10605

Signed-off-by: anjali-syna <206662904+anjali-syna@users.noreply.github.com>

anjali-syna committed:
Removing unused variable

Signed-off-by: anjali-syna <206662904+anjali-syna@users.noreply.github.com>

anjali-syna committed:
Clearing mMediaHeaders

Signed-off-by: anjali-syna <206662904+anjali-syna@users.noreply.github.com>

anjali-syna committed:
Adding microtests

Signed-off-by: anjali-syna <206662904+anjali-syna@users.noreply.github.com>

anjali-syna committed:
Merge branch 'dev_sprint_25_2' into feature/VPLAY-10605

anjali-syna committed:
Merge branch 'dev_sprint_25_2' into feature/VPLAY-10605

anjali-syna committed:
Adding one more microtest

Signed-off-by: anjali-syna <206662904+anjali-syna@users.noreply.github.com>

anjali-syna committed:
Merge branch 'feature/VPLAY-10605' of github.com:rdkcentral/aamp into feature/VPLAY-10605

p-bond commented on AampStreamSinkManager.cpp:
```diff
@@ -616,3 +620,41 @@ void AampStreamSinkManager::UpdateTuningPlayer(PrivateInstanceAAMP *aamp)
+void AampStreamSinkManager::AddMediaHeader(int track, std::shared_ptr<AampStreamSinkManager::MediaHeader> header)
+{
+	std::lock_guard<std::mutex> lock(mStreamSinkMutex);
+	AAMPLOG_INFO("Entry for track = %d", track);
```
Good, but as WARN/INFO is sent on both paths (with the track #), is it superfluous?

anjali-syna replied:
Removed

p-bond commented on AampStreamSinkManager.cpp:
```diff
@@ -616,3 +620,41 @@ void AampStreamSinkManager::UpdateTuningPlayer(PrivateInstanceAAMP *aamp)
+	{
+		AAMPLOG_INFO("AampStreamSinkManager(%p) track[%d] url[%s] mimeType[%s] injected[%d]", 
+			this, track, it->second->url.c_str(), it->second->mimeType.c_str(), it->second->injected);
+		return it->second;
```
Comcast coding guidelines advocate single return path.
Which may or may not adversely impact code readability.

anjali-syna replied:
done

p-bond commented on AampStreamSinkManager.cpp:
```diff
@@ -616,3 +620,41 @@ void AampStreamSinkManager::UpdateTuningPlayer(PrivateInstanceAAMP *aamp)
+			this, track, it->second->url.c_str(), it->second->mimeType.c_str(), it->second->injected);
+		return it->second;
+	}
+	return nullptr;
```
Prefer return {} for an empty shared_ptr.

anjali-syna replied:
This code is modified.

p-bond commented on AampStreamSinkManager.h:
```diff
@@ -178,12 +213,15 @@ class AampStreamSinkManager
 	std::map<PrivateInstanceAAMP*, AampStreamSinkInactive*> mInactiveGstPlayersMap;			/**< To maintain information on currently inactive PrivateInstanceAAMP*/
 	std::map<int, std::string> mEncryptedHeaders;
 
+	std::map<int, std::shared_ptr<MediaHeader>> mMediaHeaders;
```
Can track be < 0?  Assuming not, in which case unsigned.  Goes for the previous line too (not part of this delta, but fits with Comcast's guidelines)
Per Dave's comments (sort of), does this need to be a map or could it be a vector?

anjali-syna replied:
moved to vector as discussed.

dnfaulkner commented on AampStreamSinkManager.h:
```diff
@@ -178,12 +213,15 @@ class AampStreamSinkManager
 	std::map<PrivateInstanceAAMP*, AampStreamSinkInactive*> mInactiveGstPlayersMap;			/**< To maintain information on currently inactive PrivateInstanceAAMP*/
 	std::map<int, std::string> mEncryptedHeaders;
 
+	std::map<int, std::shared_ptr<MediaHeader>> mMediaHeaders;
```
Why use of map? Why not just define an array:
std::shared_ptr<MediaHeader> mMediaHeaders[AAMP_TRACK_COUNT];

anjali-syna replied:
moved to vector as discussed.

dnfaulkner commented on AampStreamSinkManager.cpp:
```diff
@@ -616,3 +620,41 @@ void AampStreamSinkManager::UpdateTuningPlayer(PrivateInstanceAAMP *aamp)
+std::shared_ptr<AampStreamSinkManager::AampStreamSinkManager::MediaHeader> AampStreamSinkManager::GetMediaHeader(int track)
+{
+	std::lock_guard<std::mutex> lock(mStreamSinkMutex);
+	auto it = mMediaHeaders.find(track);
```
if mMediaHeaders was an array, then you could do:
return mMediaHeaders[track];

anjali-syna replied:
done

dnfaulkner commented on AampStreamSinkManager.cpp:
```diff
@@ -616,3 +620,41 @@ void AampStreamSinkManager::UpdateTuningPlayer(PrivateInstanceAAMP *aamp)
+	std::lock_guard<std::mutex> lock(mStreamSinkMutex);
+	AAMPLOG_INFO("Entry for track = %d", track);
+
+	if (mMediaHeaders.count(track))
```
if mMediaHeaders was an array, then you could do:
if (mMediaHeaders[track])

p-bond replied:
Same with a vector.

anjali-syna replied:
done

dnfaulkner commented on AampStreamSinkManager.cpp:
```diff
@@ -77,6 +78,7 @@ void AampStreamSinkManager::Clear(void)
 	mPipelineMode = ePIPELINEMODE_UNDEFINED;
 	mEncryptedHeaders.clear();
 	mEncryptedHeadersInjected = false;
+	mMediaHeaders.clear();
```
if mMediaHeaders was an array, then you could declare a function to clear the array:
for (track = 0; track < AAMP_TRACK_COUNT; track++) mMediaHeaders[track] = nullptr;

p-bond replied:
Would strongly advocate for STL types.  Vector possibly, but not an array please.

anjali-syna replied:
Changed to vector as discussed.

dnfaulkner commented on AampStreamSinkManager.cpp:
```diff
@@ -245,6 +247,7 @@ void AampStreamSinkManager::DeleteStreamSink(PrivateInstanceAAMP *aamp)
 				mGstPlayer = nullptr;
 				mPipelineMode = ePIPELINEMODE_UNDEFINED;
 				mEncryptedHeadersInjected = false;
+				mMediaHeaders.clear();
```
See line 81

anjali-syna replied:
using vector as discussed

dnfaulkner commented on AampStreamSinkManager.cpp:
```diff
@@ -356,6 +359,7 @@ void AampStreamSinkManager::DeactivatePlayer(PrivateInstanceAAMP *aamp, bool sto
 					AAMPLOG_WARN("AampStreamSinkManager(%p) Single Pipeline mode, deactivating and stopping active PLAYER[%d]", this, aamp->mPlayerId);
 					mEncryptedHeadersInjected = false;
 					mEncryptedHeaders.clear();
+					mMediaHeaders.clear();
```
See line 81

anjali-syna replied:
using vector as discussed

dnfaulkner commented on AampStreamSinkManager.cpp:
```diff
@@ -616,3 +620,41 @@ void AampStreamSinkManager::UpdateTuningPlayer(PrivateInstanceAAMP *aamp)
 	}
 }
+
+void AampStreamSinkManager::AddMediaHeader(int track, std::shared_ptr<AampStreamSinkManager::MediaHeader> header)
```
Is this functionality actually used for anything other than subtitles? What are the use cases?

The method prototype signature in the header file is:
void AddMediaHeader(int track, std::shared_ptr<**MediaHeader**> header);
Why is the method signature different:
void AampStreamSinkManager::AddMediaHeader(int track, std::shared_ptr<**AampStreamSinkManager::MediaHeader**> header)

anjali-syna replied:
This is because MediaHeader class is defined inside class AampStreamSinkManager

dnfaulkner commented on AampStreamSinkManager.cpp:
```diff
@@ -616,3 +620,41 @@ void AampStreamSinkManager::UpdateTuningPlayer(PrivateInstanceAAMP *aamp)
+	AAMPLOG_INFO("AampStreamSinkManager(%p) Removed header for track[%d]", this, track);
+}
+
+std::shared_ptr<AampStreamSinkManager::AampStreamSinkManager::MediaHeader> AampStreamSinkManager::GetMediaHeader(int track)
```
The method prototype signature in the header file is:
std::shared_ptr<**MediaHeader**> GetMediaHeader(int track);
Why is the method signature different:
std::shared_ptr<**AampStreamSinkManager::AampStreamSinkManager::MediaHeader**> AampStreamSinkManager::GetMediaHeader(int track)
?

anjali-syna replied:
Removed the additional AampStreamSinkManager::

p-bond commented on fragmentcollector_mpd.cpp:
```diff
@@ -8984,6 +8994,91 @@ bool StreamAbstractionAAMP_MPD::GetEncryptedHeaders(std::map<int, std::string>&
+		if(period != NULL)
+		{
+			size_t numAdaptationSets = period->GetAdaptationSets().size();
+			for(unsigned iAdaptationSet = 0; iAdaptationSet < numAdaptationSets && !subtitleFound; iAdaptationSet++)
```
As GetAdaptationSets() returns a vector, this could be a range-based for instead.
Obviously would need to break if it finds a subtitle.
Cleaner code.

anjali-syna replied:
done

dnfaulkner commented on AampStreamSinkManager.cpp:
```diff
@@ -616,3 +620,41 @@ void AampStreamSinkManager::UpdateTuningPlayer(PrivateInstanceAAMP *aamp)
+{
+	std::lock_guard<std::mutex> lock(mStreamSinkMutex);
+	auto it = mMediaHeaders.find(track);
+	if (it != mMediaHeaders.end())
```
If mMediaHeaders was an array, then you could do:
if (mMediaHeaders[track])
{
    AAMPLOG_INFO("AampStreamSinkManager(%p) track[%d] url[%s] mimeType[%s] injected[%d]", 
			this, track, mMediaHeaders->url.c_str(), mMediaHeaders->mimeType.c_str(), mMediaHeaders->injected);
} else {
    AAMPLOG_INFO("AampStreamSinkManager(%p) track[%d] has not been added");
}
return mMediaHeaders[track];

anjali-syna replied:
modified

p-bond commented on fragmentcollector_mpd.cpp:
```diff
@@ -4291,6 +4291,16 @@ AAMPStatusType StreamAbstractionAAMP_MPD::Init(TuneType tuneType)
+		// later if a pre-roll advert is played that does not contain subtitles.
+		if (ISCONFIGSET(eAAMPConfig_useRialtoSink) && 
+		   !mIsLiveStream &&
+		   (nullptr == AampStreamSinkManager::GetInstance().GetMediaHeader(eMEDIATYPE_SUBTITLE)))
```
Testing a shared_ptr is implicitly boolean - no need to compare with nullptr.
Idiomatic.

anjali-syna replied:
done

dnfaulkner commented on AampStreamSinkManager.cpp:
```diff
@@ -616,3 +620,41 @@ void AampStreamSinkManager::UpdateTuningPlayer(PrivateInstanceAAMP *aamp)
+		AAMPLOG_INFO("AampStreamSinkManager(%p) track[%d] url[%s] mimeType[%s] injected[%d]", 
+			this, track, it->second->url.c_str(), it->second->mimeType.c_str(), it->second->injected);
+		return it->second;
+	}
```
else
 log that it has not been added / is NULL

anjali-syna replied:
added

p-bond commented on fragmentcollector_mpd.cpp:
```diff
@@ -8984,6 +8994,91 @@ bool StreamAbstractionAAMP_MPD::GetEncryptedHeaders(std::map<int, std::string>&
+
+	while(iPeriod < numPeriods && !subtitleFound)
+	{
+		IPeriod *period = mpd->GetPeriods().at(iPeriod);
```
GetPeriods returns a vector - please use a range-based for instead of counting through the offsets.

anjali-syna replied:
done

dnfaulkner commented on fragmentcollector_mpd.cpp:
```diff
@@ -8984,6 +8994,91 @@ bool StreamAbstractionAAMP_MPD::GetEncryptedHeaders(std::map<int, std::string>&
+	bool subtitleFound = false;
+	unsigned iPeriod = 0;
+
+	while(iPeriod < numPeriods && !subtitleFound)
```
Why isn't this:
for (iPeriod = 0; iPeriod < numPeriods && !subtitleFound; iPeriod++)
?

p-bond commented on fragmentcollector_mpd.cpp:
```diff
@@ -8984,6 +8994,91 @@ bool StreamAbstractionAAMP_MPD::GetEncryptedHeaders(std::map<int, std::string>&
+							{
+								std::string fragmentUrl;
+								FragmentDescriptor *fragmentDescriptor = new FragmentDescriptor();
+								std::shared_ptr<AampStreamSinkManager::MediaHeader> subtitleHeader = std::make_shared<AampStreamSinkManager::MediaHeader>();
```
Opportunity to use auto for readability

anjali-syna replied:
done

p-bond commented on fragmentcollector_mpd.cpp:
```diff
@@ -10847,25 +10942,44 @@ void StreamAbstractionAAMP_MPD::GetStreamFormat(StreamOutputFormat &primaryOutpu
 		{
-			AAMPLOG_INFO("mimeType empty");
-			subtitleOutputFormat = FORMAT_SUBTITLE_MP4;
+			std::shared_ptr<AampStreamSinkManager::MediaHeader> subtitleHeader = AampStreamSinkManager::GetInstance().GetMediaHeader(eMEDIATYPE_SUBTITLE);
```
auto

anjali-syna replied:
done

dnfaulkner commented on fragmentcollector_mpd.cpp:
```diff
@@ -8984,6 +8994,91 @@ bool StreamAbstractionAAMP_MPD::GetEncryptedHeaders(std::map<int, std::string>&
 	return ret;
 }
+/**
+ * @brief Find subtitle adaptationSet if available
```
I see that this only looks for the first subtitle adaptation set - not the actual one that gets selected according to configured language. Theoretically, each language or rendition may have a different coding (but probably doesn't) so you're potentially extracting the wrong header, which may have different encoding or other characteristics.
You should add a note to highlight this limitation or find some existing code which extracts the correct header according to the configured subtitles (set via SetPreferredTextLanguages()). I think that the former will be fine though.

anjali-syna replied:
comment added in the function header

p-bond commented on fragmentcollector_mpd.cpp:
```diff
@@ -11533,12 +11647,49 @@ void StreamAbstractionAAMP_MPD::StopInjection(void)
+ */
+void StreamAbstractionAAMP_MPD::SendMediaHeaders()
+{
+	for (int iTrack = 0; iTrack < mMaxTracks; iTrack++)
```
If mMediaStreamContext was a vector, this could become a range based for.
Possible refactor for idiomatic C++?

anjali-syna replied:
mMediaStreamContext is an array. Existing code; Not touching it as of now

p-bond commented on fragmentcollector_mpd.cpp:
```diff
@@ -11533,12 +11647,49 @@ void StreamAbstractionAAMP_MPD::StopInjection(void)
+		MediaStreamContext *track = mMediaStreamContext[iTrack];
+		if(track && !track->Enabled())
+		{
+			std::shared_ptr<AampStreamSinkManager::MediaHeader> header = AampStreamSinkManager::GetInstance().GetMediaHeader(iTrack);
```
auto...

anjali-syna replied:
done

p-bond commented on fragmentcollector_mpd.cpp:
```diff
@@ -11533,12 +11647,49 @@ void StreamAbstractionAAMP_MPD::StopInjection(void)
+				AAMPLOG_INFO("Track is disabled; url for init segment found: %s", header->url.c_str());
+				AampGrowableBuffer buffer("init-buffer");
+				std::string effectiveUrl;
+				int http_error = 0;
```
Please use universal initialisation in preference.

anjali-syna replied:
done

p-bond reviewed (commented):
Please address the Coverity warnings too - they all seem valid.

dnfaulkner commented on fragmentcollector_mpd.cpp:
```diff
@@ -11533,12 +11647,49 @@ void StreamAbstractionAAMP_MPD::StopInjection(void)
 }
 
+/**
+ * @brief Send any cached init fragments to be injected on disabled streams to generate the pipeline
```
Did we really need a generic implementation to handle any disabled streams rather than specifically just subtitles? What are the use cases?

anjali-syna replied:
probably subtitles are the only one as of now

dnfaulkner reviewed (commented):
Some feedback

anjali-syna committed:
Addressed review comments.

Signed-off-by: anjali-syna <206662904+anjali-syna@users.noreply.github.com>

p-bond commented on fragmentcollector_mpd.cpp:
```diff
@@ -11657,13 +11660,13 @@ void StreamAbstractionAAMP_MPD::SendMediaHeaders()
 				AampGrowableBuffer buffer("init-buffer");
 				std::string effectiveUrl;
-				int http_error = 0;
+				int http_error = {};
```
Can also use
int http_error() or int http_error{}
Strictly the 2nd variant is an initialiser list in C++11, but seems to work.
https://en.cppreference.com/w/cpp/language/initialization.html

anjali-syna replied:
done

p-bond commented on test/utests/drm/mocks/FakeAampStreamSinkManager.cpp:
```diff
@@ -97,15 +97,15 @@ void AampStreamSinkManager::UpdateTuningPlayer(PrivateInstanceAAMP *aamp)
-std::shared_ptr<AampStreamSinkManager::MediaHeader> AampStreamSinkManager::GetMediaHeader(int track)
+std::shared_ptr<AampStreamSinkManager::MediaHeader> AampStreamSinkManager::GetMediaHeader(unsigned track)
 {
 	return nullptr;
```
return {};

anjali-syna replied:
done

p-bond commented on test/utests/fakes/FakeAampStreamSinkManager.cpp:
```diff
@@ -103,15 +103,15 @@ void AampStreamSinkManager::UpdateTuningPlayer(PrivateInstanceAAMP *aamp)
-std::shared_ptr<AampStreamSinkManager::MediaHeader> AampStreamSinkManager::GetMediaHeader(int track)
+std::shared_ptr<AampStreamSinkManager::MediaHeader> AampStreamSinkManager::GetMediaHeader(unsigned track)
 {
 	return nullptr;
```
return {};

anjali-syna replied:
done

anjali-syna committed:
Addressed new review comments

Signed-off-by: anjali-syna <206662904+anjali-syna@users.noreply.github.com>

p-bond committed:
Merge branch 'dev_sprint_25_2' into feature/VPLAY-10605

anjali-syna committed:
Modifying fakes

Signed-off-by: anjali-syna <206662904+anjali-syna@users.noreply.github.com>

anjali-syna committed:
Merge branch 'feature/VPLAY-10605' of github.com:rdkcentral/aamp into feature/VPLAY-10605

dnfaulkner reviewed (commented):
I've added some comments

dnfaulkner commented on AampStreamSinkManager.cpp:
```diff
@@ -616,3 +620,43 @@ void AampStreamSinkManager::UpdateTuningPlayer(PrivateInstanceAAMP *aamp)
+void AampStreamSinkManager::RemoveMediaHeader(unsigned track)
+{
+	std::lock_guard<std::mutex> lock(mStreamSinkMutex);
+	mMediaHeaders.erase(mMediaHeaders.begin() + track);
```
According to Pete, this moves all the elements down in the array above the one erased.
Should this be:
mMediaHeaders[track].reset();
?

p-bond replied:
Also - for future ref, not asking for a rewrite now - if there's only ever one owner, it's better to use a unique_ptr rather than shared_ptr.
Although you don't get make_unique<x> until C+14 leaving you with 
std::unique_ptr<x> y(new x())
Which is far less readable.

anjali-syna replied:
used reset

dnfaulkner commented on AampStreamSinkManager.cpp:
```diff
@@ -616,3 +620,43 @@ void AampStreamSinkManager::UpdateTuningPlayer(PrivateInstanceAAMP *aamp)
+	}
+	else
+	{
+		AAMPLOG_WARN("AampStreamSinkManager(%p) unable to find MediaHeaders for track[%u]", this, track);
```
Replace "MediaHeaders" by "MediaHeader"

anjali-syna replied:
done

dnfaulkner commented on fragmentcollector_mpd.cpp:
```diff
@@ -8984,6 +8994,94 @@ bool StreamAbstractionAAMP_MPD::GetEncryptedHeaders(std::map<int, std::string>&
 }
+/**
+ * @brief Find subtitle adaptationSet if available. Note - Currently
+ * this function looks only for the first subtitle adaptation set - not the one
```
Add:
this function assumes that all subtitle adaptation sets use the same mimeType / codec and there are no significant differences in the header fragments, so it looks only for the first

anjali-syna replied:
done

dnfaulkner commented on fragmentcollector_mpd.cpp:
```diff
@@ -8984,6 +8994,94 @@ bool StreamAbstractionAAMP_MPD::GetEncryptedHeaders(std::map<int, std::string>&
+/**
+ * @brief Find subtitle adaptationSet if available. Note - Currently
+ * this function looks only for the first subtitle adaptation set - not the one
+ * that gets selected according to configured language.
```
Add:
or other criteria

anjali-syna replied:
done

dnfaulkner commented on fragmentcollector_mpd.cpp:
```diff
@@ -8984,6 +8994,94 @@ bool StreamAbstractionAAMP_MPD::GetEncryptedHeaders(std::map<int, std::string>&
+
+	for (auto &period: mpd->GetPeriods())
+	{
+		if(period != NULL)
```
How is NULL possible now that we have "for (auto &period: mpd->GetPeriods())"?

p-bond replied:
(Also, if period is a raw pointer, use nullptr; if it's a smart ptr use it as a boolean)

anjali-syna replied:
used nullptr as suggested by Peter

dnfaulkner commented on fragmentcollector_mpd.cpp:
```diff
@@ -8984,6 +8994,94 @@ bool StreamAbstractionAAMP_MPD::GetEncryptedHeaders(std::map<int, std::string>&
+		{
+			for(auto &adaptationSet: period->GetAdaptationSets())
+			{
+				if(adaptationSet != NULL)
```
How is NULL possible now that we have for(auto &adaptationSet: period->GetAdaptationSets())?

anjali-syna replied:
used nullptr as suggested by Peter

dnfaulkner reviewed (commented):
6 comments

anjali-syna committed:
Addressing new Review comments

Signed-off-by: anjali-syna <206662904+anjali-syna@users.noreply.github.com>

anjali-syna committed:
Additional logging

Signed-off-by: anjali-syna <206662904+anjali-syna@users.noreply.github.com>

anjali-syna committed:
Additional checks added

Signed-off-by: anjali-syna <206662904+anjali-syna@users.noreply.github.com>

anjali-syna committed:
Modified condition check

Signed-off-by: anjali-syna <206662904+anjali-syna@users.noreply.github.com>

anjali-syna committed:
Removed commented code

Signed-off-by: anjali-syna <206662904+anjali-syna@users.noreply.github.com>

anjali-syna committed:
Removed commented code

Signed-off-by: anjali-syna <206662904+anjali-syna@users.noreply.github.com>

anjali-syna committed:
Removed commented code

Signed-off-by: anjali-syna <206662904+anjali-syna@users.noreply.github.com>

anjali-syna committed:
Merge branch 'dev_sprint_25_2' into feature/VPLAY-10605

anjali-syna committed:
Merge branch 'dev_sprint_25_2' into feature/VPLAY-10605
